rust/src/state_machine/mod.rs

```diff
@@ -0,0 +1,156 @@
+    CoordinatorPublicKey,
+    CoordinatorSecretKey,
+    InitError,
+    PetError,
+    SeedDict,
+    SumDict,
+};
+use std::default::Default;
+use thiserror::Error;
+use tokio::sync::{mpsc, oneshot};
+
+mod error;
+mod idle;
+pub mod requests;
+mod sum;
+mod sum2;
+mod update;
+
+use requests::Request;
+
+#[derive(Debug, Clone)]
+pub struct CoordinatorState {
+    pk: CoordinatorPublicKey, // 32 bytes
+    sk: CoordinatorSecretKey, // 32 bytes
+
+    // round parameters
+    sum: f64,
+    update: f64,
+    seed: RoundSeed,
+    min_sum: usize,
+    min_update: usize,
+
+    // round dictionaries
+    /// Dictionary built during the sum phase.
+    sum_dict: SumDict,
+    /// Dictionary built during the update phase.
+    seed_dict: SeedDict,
+    /// Dictionary built during the sum2 phase.
+    mask_dict: MaskDict,
+
+    /// The masking configuration
+    mask_config: MaskConfig,
+}
+
+impl Default for CoordinatorState {
+    fn default() -> Self {
+        let pk = CoordinatorPublicKey::zeroed();
+        let sk = CoordinatorSecretKey::zeroed();
+        let sum = 0.4_f64;
+        let update = 0.5_f64;
+        let seed = RoundSeed::zeroed();
+        let min_sum = 1_usize;
+        let min_update = 3_usize;
+        let sum_dict = SumDict::new();
+        let seed_dict = SeedDict::new();
+        let mask_dict = MaskDict::new();
+        let mask_config = MaskConfig {
+            group_type: GroupType::Prime,
+            data_type: DataType::F32,
+            bound_type: BoundType::B0,
+            model_type: ModelType::M3,
+        };
+        Self {
+            pk,
+            sk,
+            sum,
+            update,
+            seed,
+            min_sum,
+            min_update,
+            sum_dict,
+            seed_dict,
+            mask_dict,
+            mask_config,
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum StateError {
+    #[error("state failed: channel error: {0}")]
+    ChannelError(&'static str),
+    #[error("state failed: round error: {0}")]
+    RoundError(#[from] RoundFailed),
+}
+
+pub struct State<S> {
+    // Inner state
+    _inner: S,
+    // Coordinator state
+    coordinator_state: CoordinatorState,
+    // Request receiver
+    request_rx: mpsc::UnboundedReceiver<Request>,
+}
+
+// Functions that are available to all states
+impl<S> State<S> {
+    /// Receives the next [`Request`].
+    /// Returns [`StateError::ChannelError`] when all senders are dropped.
+    async fn next_request(&mut self) -> Result<Request, StateError> {
+        debug!("received new message");
+        self.request_rx.recv().await.ok_or(StateError::ChannelError(
+            "all message senders have been dropped!",
+        ))
+    }
+
+    // Handle invalid requests.
+    fn handle_invalid_message(response_tx: oneshot::Sender<Result<(), PetError>>) {
+        debug!("invalid message");
+        let _ = response_tx.send(Err(PetError::InvalidMessage));
```

Comment:
It's the other way around. When I remove it, cargo throws a warning

```
warning: unused `std::result::Result` that must be used
   --> src/state_machine/mod.rs:120:9
    |
120 |         response_tx.send(Err(PetError::InvalidMessage));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: this `Result` may be an `Err` variant, which should be handled

warning: 1 warning emitted
```

Reply:
Yes you do need `let _` to make it clear that you don't care about a `Result`, because it has the `#[must_use]` attribute.

Reply:
ah i see, i didn't notice that the return type is a `Result<>`. then we should probably handle the result appropriately when error handling will be improved.

Reply:
I think in this case it's fine to not handle the error. The only reason this would error out is if the receiver was dropped, which just means that whoever made the request was not interested in the response. This is probably worth a comment though.

Reply:
I have added a comment